File: src/main/java/com/github/unchama/targetedeffect/UnfocusedEffect.kt
package com.github.unchama.targetedeffect

class UnfocusedEffect(val effect: suspend (Any?) -> Unit): TargetedEffect<Any?> {
  override suspend fun runFor(minecraftObject: Any?) = effect(minecraftObject)
}

/**
 * 副作用を持つ一般の[effect]を[UnfocusedEffect]として扱えるように変換する.
 */
fun unfocusedEffect(effect: suspend () -> Unit): UnfocusedEffect = UnfocusedEffect { effect() }

File: src/main/java/com/github/unchama/targetedeffect/UnfocusedEffect.scala
package com.github.unchama.targetedeffect

import com.github.unchama.util.kotlin2scala.SuspendingMethod
import kotlin.coroutines.Continuation

case class UnfocusedEffect(effect: Continuation[Unit] => Any => Unit) extends TargetedEffect[Any] {
  override @SuspendingMethod def runFor(minecraftObject: Any) = effect(continuation)(minecraftObject)
}
